app/main.py

```diff
@@ -86,129 +86,59 @@
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
-
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
-
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
-
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
-
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
-
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
-
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
-
-    # Return battle results:
-    return {
-        lancelot["name"]: lancelot["hp"],
-        arthur["name"]: arthur["hp"],
-        mordred["name"]: mordred["hp"],
-        red_knight["name"]: red_knight["hp"],
-    }
+class Knight:
+
+    knights = {}
+
+    def __init__(self, name: str, power, hp, protection):
+        self.name = name
+        self.power = power
+        self.hp = hp
+        self.protection = protection
+        Knight.knights[self.name] = self
+
+    def __sub__(self, other):
+        self.hp -= other.power - self.protection
+        if self.hp < 0:
+            self.hp = 0
+        other.hp -= self.power - other.protection
+        if other.hp < 0:
+            other.hp = 0
+
+
+def knight_to_class(knights):
+    fighters = [value for key, value in knights.items()]
+    for knight in fighters:
+        name = knight["name"]
+        power = knight["power"] + knight["weapon"]["power"]
+        hp = knight["hp"]
+        protection = sum(i["protection"] for i in knight["armour"])
+
+        if knight["potion"] is not None:
+            if "power" in knight["potion"]["effect"]:
+                power += knight["potion"]["effect"]["power"]
+            if "hp" in knight["potion"]["effect"]:
+                hp += knight["potion"]["effect"]["hp"]
+            if "protection" in knight["potion"]["effect"]:
+                protection += knight["potion"]["effect"]["protection"]
```

Comment:
These code duplications can be fixed, using `stats = ("protection", "power", "hp")` and looping on `stats`